In [1]:
## Import Packages:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import theano
import theano.tensor as T
import keras
from keras import backend as K
from keras import initializers
from keras.regularizers import l1, l2, l1_l2
from keras.models import Sequential, Model
from keras.layers.core import Dense, Lambda, Activation
from keras.layers import Embedding, Input, Dense, Concatenate, Reshape, Multiply, Flatten, Dropout
from keras.optimizers import Adagrad, Adam, SGD, RMSprop
from evaluate import evaluate_model
from Dataset import Dataset
from time import time
import sys
import GMF, MLP
import argparse

Using TensorFlow backend.


In [ ]:
'''
Created on Aug 9, 2016
Keras Implementation of Neural Matrix Factorization (NeuMF) recommender model in:
He Xiangnan et al. Neural Collaborative Filtering. In WWW 2017.  

@author: Xiangnan He (xiangnanhe@gmail.com)
'''
#################### Arguments ####################
def parse_args():
    parser = argparse.ArgumentParser(description="Run NeuMF.")
    parser.add_argument('--path', nargs='?', default='Data/',
                        help='Input data path.')
    parser.add_argument('--dataset', nargs='?', default='ml-1m',
                        help='Choose a dataset.')
    parser.add_argument('--epochs', type=int, default=100,
                        help='Number of epochs.')
    parser.add_argument('--batch_size', type=int, default=256,
                        help='Batch size.')
    parser.add_argument('--num_factors', type=int, default=8,
                        help='Embedding size of MF model.')
    parser.add_argument('--layers', nargs='?', default='[64,32,16,8]',
                        help="MLP layers. Note that the first layer is the concatenation of user and item embeddings. So layers[0]/2 is the embedding size.")
    parser.add_argument('--reg_mf', type=float, default=0,
                        help='Regularization for MF embeddings.')                    
    parser.add_argument('--reg_layers', nargs='?', default='[0,0,0,0]',
                        help="Regularization for each MLP layer. reg_layers[0] is the regularization for embeddings.")
    parser.add_argument('--num_neg', type=int, default=4,
                        help='Number of negative instances to pair with a positive instance.')
    parser.add_argument('--lr', type=float, default=0.001,
                        help='Learning rate.')
    parser.add_argument('--learner', nargs='?', default='adam',
                        help='Specify an optimizer: adagrad, adam, rmsprop, sgd')
    parser.add_argument('--verbose', type=int, default=1,
                        help='Show performance per X iterations')
    parser.add_argument('--out', type=int, default=1,
                        help='Whether to save the trained model.')
    parser.add_argument('--mf_pretrain', nargs='?', default='',
                        help='Specify the pretrain model file for MF part. If empty, no pretrain will be used')
    parser.add_argument('--mlp_pretrain', nargs='?', default='',
                        help='Specify the pretrain model file for MLP part. If empty, no pretrain will be used')
    return parser.parse_args()

def get_model(num_users, num_items, mf_dim=10, layers=[10], reg_layers=[0], reg_mf=0):
    assert len(layers) == len(reg_layers)
    num_layer = len(layers) #Number of layers in the MLP
    # Input variables
    user_input = Input(shape=(1,), dtype='int32', name = 'user_input')
    item_input = Input(shape=(1,), dtype='int32', name = 'item_input')
    
    print('Input layer: ', user_input.shape, user_input.dtype)
    
    # Embedding layer
    MF_Embedding_User = Embedding(input_dim = num_users, output_dim = mf_dim, name = 'mf_embedding_user',
                                  embeddings_initializer = 'random_normal', W_regularizer = l2(reg_mf), input_length=1)
    MF_Embedding_Item = Embedding(input_dim = num_items, output_dim = mf_dim, name = 'mf_embedding_item',
                                  embeddings_initializer = 'random_normal', W_regularizer = l2(reg_mf), input_length=1)  

    MLP_Embedding_User = Embedding(input_dim = num_users, output_dim = int(layers[0]/2), name = "mlp_embedding_user",
                                   embeddings_initializer = 'random_normal', W_regularizer = l2(reg_layers[0]), input_length=1)
    MLP_Embedding_Item = Embedding(input_dim = num_items, output_dim = int(layers[0]/2), name = 'mlp_embedding_item',
                                   embeddings_initializer = 'random_normal', W_regularizer = l2(reg_layers[0]), input_length=1)

    output = MLP_Embedding_User(user_input)
    #print(MLP_Embedding_User.weights)
    
    # MF part
    mf_user_latent = Flatten()(MF_Embedding_User(user_input))
    mf_item_latent = Flatten()(MF_Embedding_Item(item_input))
    
    mf_vector = Multiply()([mf_user_latent, mf_item_latent]) # element-wise multiply

    # MLP part 
    mlp_user_latent = Flatten()(MLP_Embedding_User(user_input))
    mlp_item_latent = Flatten()(MLP_Embedding_Item(item_input))
    
    mlp_vector = Concatenate()([mlp_user_latent, mlp_item_latent])
    for idx in range(1, num_layer):
        layer = Dense(layers[idx], W_regularizer= l2(reg_layers[idx]), activation='relu', name="layer%d" %idx)
        mlp_vector = layer(mlp_vector)

    # Concatenate MF and MLP parts
    #mf_vector = Lambda(lambda x: x * alpha)(mf_vector)
    #mlp_vector = Lambda(lambda x : x * (1-alpha))(mlp_vector)
    predict_vector = Concatenate()([mf_vector, mlp_vector])
    
    # Final prediction layer
    prediction = Dense(1, activation='sigmoid', init='lecun_uniform', name = "prediction")(predict_vector)
    
    model = Model(input=[user_input, item_input], 
                  output=prediction)
    
    return model

def load_pretrain_model(model, gmf_model, mlp_model, num_layers):
    # MF embeddings
    gmf_user_embeddings = gmf_model.get_layer('user_embedding').get_weights()
    gmf_item_embeddings = gmf_model.get_layer('item_embedding').get_weights()
    model.get_layer('mf_embedding_user').set_weights(gmf_user_embeddings)
    model.get_layer('mf_embedding_item').set_weights(gmf_item_embeddings)
    
    # MLP embeddings
    mlp_user_embeddings = mlp_model.get_layer('user_embedding').get_weights()
    mlp_item_embeddings = mlp_model.get_layer('item_embedding').get_weights()
    model.get_layer('mlp_embedding_user').set_weights(mlp_user_embeddings)
    model.get_layer('mlp_embedding_item').set_weights(mlp_item_embeddings)
    
    # MLP layers
    for i in range(1, num_layers):
        mlp_layer_weights = mlp_model.get_layer('layer%d' %i).get_weights()
        model.get_layer('layer%d' %i).set_weights(mlp_layer_weights)
        
    # Prediction weights
    gmf_prediction = gmf_model.get_layer('prediction').get_weights()
    mlp_prediction = mlp_model.get_layer('prediction').get_weights()
    new_weights = np.concatenate((gmf_prediction[0], mlp_prediction[0]), axis=0)
    new_b = gmf_prediction[1] + mlp_prediction[1]
    model.get_layer('prediction').set_weights([0.5*new_weights, 0.5*new_b])    
    return model

def get_train_instances(train, num_negatives):
    user_input, item_input, labels = [],[],[]
    num_users = train.shape[0]
    s1 = set(train.keys())
    
    for (u, i) in train.keys():
        # positive instance
        user_input.append(u)
        item_input.append(i)
        labels.append(1)
        # negative instances
        for t in range(num_negatives):
            j = np.random.randint(num_items)
            while (u, j) in s1:
                j = np.random.randint(num_items)
            user_input.append(u)
            item_input.append(j)
            labels.append(0)
    return user_input, item_input, labels

In [ ]:
sys.argv = ['NeuMF.py --dataset ml-1m --epochs 20 --batch_size 256 --num_factors 8 --layers [64,32,16,8] --reg_mf 0 --reg_layers [0,0,0,0] --num_neg 4 --lr 0.001 --learner adam --verbose 1 --out 1']
#sys.argv = ['NeuMF.py --dataset ml-1m --epochs 20 --batch_size 256 --num_factors 8 --layers [64,32,16,8] --num_neg 4 --lr 0.001 --learner adam --verbose 1 --out 1 --mf_pretrain Pretrain/ml-1m_GMF_8_1501651698.h5 --mlp_pretrain Pretrain/ml-1m_MLP_[64,32,16,8]_1501652038.h5']

In [ ]:
args = parse_args()
num_epochs = args.epochs
batch_size = args.batch_size
mf_dim = args.num_factors
layers = eval(args.layers)
reg_mf = args.reg_mf
reg_layers = eval(args.reg_layers)
num_negatives = args.num_neg
learning_rate = args.lr
learner = args.learner
verbose = args.verbose
mf_pretrain = args.mf_pretrain
mlp_pretrain = args.mlp_pretrain

topK = 10
evaluation_threads = 1 # mp.cpu_count()
print("NeuMF arguments: %s " %(args))
model_out_file = 'Pretrain/%s_NeuMF_%d_%s_%d.h5' %(args.dataset, mf_dim, args.layers, time())

In [ ]:
# Loading data
t1 = time()
dataset = Dataset(args.path + args.dataset)
train, testRatings, testNegatives = dataset.trainMatrix, dataset.testRatings, dataset.testNegatives
num_users, num_items = train.shape
print("Load data done [%.1f s]. #user=%d, #item=%d, #train=%d, #test=%d" 
      %(time()-t1, num_users, num_items, train.nnz, len(testRatings)))

In [ ]:
print(num_users, num_items, mf_dim, layers, reg_layers, reg_mf)
print(type(layers[1]))

In [ ]:
# Build model
model = get_model(num_users, num_items, mf_dim, layers, reg_layers, reg_mf)

In [ ]:
if learner.lower() == "adagrad": 
    model.compile(optimizer=Adagrad(lr=learning_rate), loss='binary_crossentropy')
elif learner.lower() == "rmsprop":
    model.compile(optimizer=RMSprop(lr=learning_rate), loss='binary_crossentropy')
elif learner.lower() == "adam":
    model.compile(optimizer=Adam(lr=learning_rate), loss='binary_crossentropy')
else:
    model.compile(optimizer=SGD(lr=learning_rate), loss='binary_crossentropy')

In [ ]:
# Load pretrain model
if mf_pretrain != '' and mlp_pretrain != '':
    gmf_model = GMF.get_model(num_users,num_items,mf_dim)
    gmf_model.load_weights(mf_pretrain)
    mlp_model = MLP.get_model(num_users,num_items, layers, reg_layers)
    mlp_model.load_weights(mlp_pretrain)
    model = load_pretrain_model(model, gmf_model, mlp_model, len(layers))
    print("Load pretrained GMF (%s) and MLP (%s) models done. " %(mf_pretrain, mlp_pretrain))

In [ ]:
# Init performance
(hits, ndcgs) = evaluate_model(model, testRatings, testNegatives, topK, evaluation_threads)
hr, ndcg = np.array(hits).mean(), np.array(ndcgs).mean()
print('Init: HR = %.4f, NDCG = %.4f' % (hr, ndcg))
best_hr, best_ndcg, best_iter = hr, ndcg, -1
if args.out > 0:
    model.save_weights(model_out_file, overwrite=True) 

In [ ]:
# Training model
for epoch in range(num_epochs):
    t1 = time()
    # Generate training instances
    user_input, item_input, labels = get_train_instances(train, num_negatives)
    print('Finish generating')
    
    # Training
    hist = model.fit([np.array(user_input), np.array(item_input)], #input
                     np.array(labels), # labels 
                     batch_size=batch_size, nb_epoch=1, verbose=0, shuffle=True)
    t2 = time()

    # Evaluation
    if epoch % verbose == 0:
        (hits, ndcgs) = evaluate_model(model, testRatings, testNegatives, topK, evaluation_threads)
        hr, ndcg, loss = np.array(hits).mean(), np.array(ndcgs).mean(), hist.history['loss'][0]
        print('Iteration %d [%.1f s]: HR = %.4f, NDCG = %.4f, loss = %.4f [%.1f s]' 
              % (epoch,  t2-t1, hr, ndcg, loss, time()-t2))
        if hr > best_hr:
            best_hr, best_ndcg, best_iter = hr, ndcg, epoch
            if args.out > 0:
                model.save_weights(model_out_file, overwrite=True)

print("End. Best Iteration %d:  HR = %.4f, NDCG = %.4f. " %(best_iter, best_hr, best_ndcg))
if args.out > 0:
    print("The best NeuMF model is saved to %s" %(model_out_file))

### Read in Champaign user-item interaction csv file to generate dataset:

In [2]:
champaign_user_item_df = pd.read_csv('../yelp_dataset/champaign_user_item.csv', index_col = 0)
champaign_user_item_df.head(3)

business_id                         name         address  \
0  4hWDMVtfnpyY72_5QMbthA  Old Time Meat & Deli Shoppe  2018 S Neil St   
1  4hWDMVtfnpyY72_5QMbthA  Old Time Meat & Deli Shoppe  2018 S Neil St   
2  4hWDMVtfnpyY72_5QMbthA  Old Time Meat & Deli Shoppe  2018 S Neil St   

               city state postal_code   latitude  longitude  \
0  urbana-champaign    IL       61820  40.091901 -88.245362   
1  urbana-champaign    IL       61820  40.091901 -88.245362   
2  urbana-champaign    IL       61820  40.091901 -88.245362   

   stars_business_avg  review_count  ...  \
0                 5.0            46  ...   
1                 5.0            46  ...   
2                 5.0            46  ...   

                                          categories  \
0  Meat Shops, Food, Restaurants, Specialty Food,...   
1  Meat Shops, Food, Restaurants, Specialty Food,...   
2  Meat Shops, Food, Restaurants, Specialty Food,...   

                                               hours               review_id  \
0  {'Monday': '10:0-18:0', 'Tuesday': '10:0-18:0'...  WruGbPJAuXAl5GP28fH1cA   
1  {'Monday': '10:0-18:0', 'Tuesday': '10:0-18:0'...  NozR-GCiMHWyhqRbW4TDuA   
2  {'Monday': '10:0-18:0', 'Tuesday': '10:0-18:0'...  wLWG8YkhTfu3zjAYGNSzfA   

                  user_id stars useful  funny  cool  \
0  5rM50rIojxWNQRv0Vmuulg   5.0      0      0     0   
1  9ZuE5YJQvNA4PvldrTgmXA   5.0      1      1     1   
2  w-NiZLoY-TetMRY8LOqMKw   5.0      1      0     2   

                                                text                 date  
0  This is a wonderful butchershop. Deli selectio...  2011-08-25 01:04:12  
1  I'm almost reluctant to give it 5 stars only s...  2016-04-12 05:07:23  
2  This is definitely the place to go for excelle...  2016-07-04 01:09:58  

[3 rows x 22 columns]

In [3]:
unique_business_id = champaign_user_item_df.business_id.unique()
mapping_business_id = {}
ctr = 0
for business_id in unique_business_id:
    mapping_business_id[business_id] = ctr
    ctr += 1
    
champaign_user_item_df['business_id_refined'] = champaign_user_item_df.business_id.map(mapping_business_id)

In [4]:
unique_user_id = champaign_user_item_df.user_id.unique()
mapping_user_id = {}
ctr = 0
for user_id in unique_user_id:
    mapping_user_id[user_id] = ctr
    ctr += 1
    
champaign_user_item_df['user_id_refined'] = champaign_user_item_df.user_id.map(mapping_user_id)

In [5]:
champaign_user_item_df[['user_id_refined', 'business_id_refined']].head(5)

user_id_refined  business_id_refined
0                0                    0
1                1                    0
2                2                    0
3                3                    0
4                4                    0

In [6]:
sum(champaign_user_item_df.groupby('user_id_refined')['name'].count() >= 10)

502

There are only 502 users who have reviewed 10 or more restaurants in champaign, hence need to expand the dataset a bit

### Read in toronto user-item interaction csv file to generate dataset:

In [7]:
toronto_user_item_df = pd.read_csv('../yelp_dataset/toronto_user_item.csv', index_col = 0)
toronto_user_item_df.head(3)

business_id           name              address     city state  \
0  g6AFW-zY0wDvBl9U82g4zg  Baretto Caffe  1262 Don Mills Road  toronto    ON   
1  g6AFW-zY0wDvBl9U82g4zg  Baretto Caffe  1262 Don Mills Road  toronto    ON   
2  g6AFW-zY0wDvBl9U82g4zg  Baretto Caffe  1262 Don Mills Road  toronto    ON   

  postal_code   latitude  longitude  stars_business_avg  review_count  ...  \
0     M3B 2W7  43.744703 -79.346468                 5.0           317  ...   
1     M3B 2W7  43.744703 -79.346468                 5.0           317  ...   
2     M3B 2W7  43.744703 -79.346468                 5.0           317  ...   

                    categories  \
0  Restaurants, Italian, Cafes   
1  Restaurants, Italian, Cafes   
2  Restaurants, Italian, Cafes   

                                               hours               review_id  \
0  {'Monday': '7:30-18:0', 'Tuesday': '7:30-18:0'...  SKBNW4QKNiclQ6mB2AQ8MQ   
1  {'Monday': '7:30-18:0', 'Tuesday': '7:30-18:0'...  0dsaJN8eljlYRCqPWN1JCQ   
2  {'Monday': '7:30-18:0', 'Tuesday': '7:30-18:0'...  aPUINDQsgifg_hSROs4TTA   

                  user_id stars useful  funny  cool  \
0  q3JSVBWICgXfO-zuLAp5fg   3.0      0      0     0   
1  0zW0RwIRwyJ6Qdirqvs5gA   5.0      0      0     0   
2  eurxcv4blzrEs7-IgLGt5w   5.0      0      0     0   

                                                text                 date  
0  The customer service is on point. The food was...  2018-10-04 10:57:11  
1  The staff and workers are really friendly and ...  2017-04-30 13:40:40  
2  This is one great cafe. A little hard to find ...  2015-03-18 22:16:23  

[3 rows x 22 columns]

In [8]:
unique_business_id = toronto_user_item_df.business_id.unique()
mapping_business_id = {}
ctr = 0
for business_id in unique_business_id:
    mapping_business_id[business_id] = ctr
    ctr += 1
    
toronto_user_item_df['business_id_refined'] = toronto_user_item_df.business_id.map(mapping_business_id)

In [9]:
unique_user_id = toronto_user_item_df.user_id.unique()
mapping_user_id = {}
ctr = 0
for user_id in unique_user_id:
    mapping_user_id[user_id] = ctr
    ctr += 1
    
toronto_user_item_df['user_id_refined'] = toronto_user_item_df.user_id.map(mapping_user_id)

In [10]:
toronto_user_item_df[['user_id_refined', 'business_id_refined']].head(5)

user_id_refined  business_id_refined
0                0                    0
1                1                    0
2                2                    0
3                3                    0
4                4                    0

In [11]:
sum(toronto_user_item_df.groupby('user_id_refined')['name'].count() >= 10)

7905

There are a substantial number of users who have reviewed at least 10 restaurants

### Filter the dataset to contain only the users who have reviewed at least 10 restaurants or more

In [12]:
grouped = toronto_user_item_df.groupby('user_id_refined')
toronto_user_item_filtered_df = grouped.filter(lambda x: x['name'].count() >= 10)

In [13]:
toronto_user_item_filtered_df.user_id_refined.nunique()

7905

In [14]:
toronto_user_item_filtered_df.business_id_refined.nunique()

8546

In [15]:
print('There are %d unique users and %d unique items in the dataset after filtering the dataset'\
%(toronto_user_item_filtered_df.user_id_refined.nunique(),toronto_user_item_filtered_df.business_id_refined.nunique()))

There are 7905 unique users and 8546 unique items in the dataset after filtering the dataset


In [19]:
### Create a dictionary of user-item interaction to serve as input
d = {} # key is (user_id, business_id) and value is 1.0 (to indicate an interaction)

for i in range(len(toronto_user_item_filtered_df)):
    d[(toronto_user_item_filtered_df.user_id_refined.iloc[i], toronto_user_item_filtered_df.business_id_refined.iloc[i])]=1.

In [20]:
d

{(3, 0): 1.0,
 (5, 0): 1.0,
 (6, 0): 1.0,
 (7, 0): 1.0,
 (10, 0): 1.0,
 (16, 0): 1.0,
 (19, 0): 1.0,
 (23, 0): 1.0,
 (24, 0): 1.0,
 (26, 0): 1.0,
 (28, 0): 1.0,
 (29, 0): 1.0,
 (30, 0): 1.0,
 (37, 0): 1.0,
 (41, 0): 1.0,
 (46, 0): 1.0,
 (51, 0): 1.0,
 (53, 0): 1.0,
 (56, 0): 1.0,
 (57, 0): 1.0,
 (62, 0): 1.0,
 (63, 0): 1.0,
 (69, 0): 1.0,
 (72, 0): 1.0,
 (79, 0): 1.0,
 (83, 0): 1.0,
 (85, 0): 1.0,
 (92, 0): 1.0,
 (95, 0): 1.0,
 (96, 0): 1.0,
 (100, 0): 1.0,
 (112, 0): 1.0,
 (115, 0): 1.0,
 (121, 0): 1.0,
 (123, 0): 1.0,
 (125, 0): 1.0,
 (130, 0): 1.0,
 (131, 0): 1.0,
 (134, 0): 1.0,
 (135, 0): 1.0,
 (139, 0): 1.0,
 (140, 0): 1.0,
 (148, 0): 1.0,
 (151, 0): 1.0,
 (152, 0): 1.0,
 (154, 0): 1.0,
 (156, 0): 1.0,
 (159, 0): 1.0,
 (161, 0): 1.0,
 (162, 0): 1.0,
 (167, 0): 1.0,
 (178, 0): 1.0,
 (180, 0): 1.0,
 (183, 0): 1.0,
 (184, 0): 1.0,
 (185, 0): 1.0,
 (186, 0): 1.0,
 (187, 0): 1.0,
 (188, 0): 1.0,
 (191, 0): 1.0,
 (194, 0): 1.0,
 (201, 0): 1.0,
 (205, 0): 1.0,
 (209, 0): 1.0,
 (214, 0):

In [21]:
f = open("Our_Data/toronto_user_item_refined.txt","w")
f.write(str(d))
f.close()